### PR TITLE
(CDAP-6168) Includes dataset classes information in artifact meta

### PR DESCRIPTION
**Pull request by chtyim**
- Move ArtifactClasses and ApplicationClass from cdap-api to cdap-proto
- Added scanning of dataset classes during artifact deployment.
- Cleanup ProgramClassLoader usage as field or return type.
  - This is to prepare for Class rewriting of Dataset classes through ClassLoader.

### CODE REVIEW

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT4243-8

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java [hunk 1]
```diff
@@ -54,19 +54,17 @@
   }
 
   /**
-   * Create a classloader that uses the artifact at the specified location to load classes, with access to
+   * Create a classloader load classes from the given directory, with access to
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java]: a classloader that loads classes?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java [hunk 2]
```diff
@@ -75,7 +73,8 @@ CloseableClassLoader createClassLoader(Location artifactLocation) throws IOExcep
       // TODO: CDAP-5613. We shouldn't always expose the Spark classes.
       programRunner = programRunnerFactory.create(ProgramType.SPARK);
       if (programRunner instanceof ProgramClassLoaderProvider) {
-        programClassLoader = ((ProgramClassLoaderProvider) programRunner).createProgramClassLoader(cConf, unpackDir);
+        programClassLoader = new ProgramClassLoader(
+          cConf, unpackDir, ((ProgramClassLoaderProvider) programRunner).createProgramClassLoaderParent());
       }
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java]: what if it is not a ProgramClassLoaderProvider? Should this throw an exception? You do expect an exception in line 81.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java (hunk 1 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java]: maybe say here: from a directory where an artifact jar has been expanded

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java [hunk 3]
```diff
@@ -54,19 +54,17 @@
...
 
   /**
-   * Create a classloader that uses the artifact at the specified location to load classes, with access to
+   * Create a classloader load classes from the given directory, with access to
    * packages that all program type has access to. The classloader created is only for artifact inspection purpose
    * and shouldn't be used for program execution as it doesn't have the proper class filtering for the specific
    * program type for the program being executed.
    *
-   * @param artifactLocation the location of the artifact to create the classloader from
+   * @param unpackDir the directory where the artifact jar is expanded to
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java]: ...jar has been expanded? That makes it clearer that the caller has to unjar.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java [hunk 4]
```diff
@@ -155,47 +212,97 @@ ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFile,
...
-        // if the user parameterized their application, like 'xyz extends Application<T>',
-        // we can deserialize the config into that object. Otherwise it'll just be a Config
-        try {
-          configType = Artifacts.getConfigType(app.getClass());
-        } catch (Exception e) {
-          throw new InvalidArtifactException(String.format(
-            "Could not resolve config type for Application class %s in artifact %s. " +
-              "The type must extend Config and cannot be parameterized.", mainClassName, artifactId));
-        }
+      Object appMain = artifactClassLoader.loadClass(mainClassName).newInstance();
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java:222]: loading the class? that is, executing user code in system process?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java [hunk 5]
```diff
@@ -501,4 +608,69 @@ public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
...
     }
   }
+
+  /**
+   * Checks if the given class extends or implements a super type.
+   *
+   * @param className name of the class to check
+   * @param superTypeName name of the super type
+   * @param resourceProvider a {@link Function} to provide {@link URL} of a class from the class name
+   * @param cache a cache for memorizing previous decision
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java]: perhaps say, previous decision for the same super type.

File: cdap-app-fabric/src/test/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeServiceTest.java [hunk 6]
```diff
@@ -72,18 +78,19 @@ public ProgramLiveInfo getLiveInfo(Id.Program programId) {
...
                                       ProgramDescriptor programDescriptor,
                                       Location programJarLocation, File tempDir) throws IOException {
         return program;
       }
+
+      @Override
+      protected ArtifactDetail getArtifactDetail(ArtifactId artifactId) throws IOException, ArtifactNotFoundException {
+        co.cask.cdap.api.artifact.ArtifactId id = new co.cask.cdap.api.artifact.ArtifactId(
+          "dummy", new ArtifactVersion("1.0"), ArtifactScope.USER);
+        return new ArtifactDetail(new ArtifactDescriptor(id, null),
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeServiceTest.java]: is this a @Nullable parameter?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java (hunk 4 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java:222]: This is not new code introduced, but just indentation changes (see Line 160 above https://github.com/caskdata/cdap/pull/6012/files/6cea330afd132113a5841be119a02af1eb348944#diff-9168b969144f6021b15078039a9dcd7aL160). Have to check with @albertshau why.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java [hunk 7]
```diff
@@ -111,29 +122,75 @@
...
         LOG.warn("Exception raised while deleting directory {}", stageDir, e);
       }
     }
-    return builder.build();
+  }
+
+  /**
+   * Extracts all dataset classes defined under the given artifact directory. It will scan every .class files under
+   * the artifact directory as well as .class files inside .jar files in the artifact directory.
+   * This method uses bytecode inspection, hence the actual class is not loaded. This is because loading every single
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java:157]: will this find dataset classes that are static inner classes of other classes?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepositoryTest.java [hunk 8]
```diff
@@ -490,7 +508,20 @@ public void testNamespaceIsolation() throws Exception {
...
-  private static ClassLoader createAppClassLoader(File jarFile) throws IOException {
+  @Test
+  public void testArtifactDataset() throws IOException, ArtifactNotFoundException {
+    ArtifactDetail artifactDetail = artifactRepository.getArtifact(APP_ARTIFACT_ID);
+
+    // The artifact being created in the setupData method is using dependency tracing
+    // hence it will pick all classes under the cdap-app-fabric/src/test/main
+    Set<DatasetClass> expectedDatasets = ImmutableSet.of(
+      new DatasetClass(TrackingTable.class.getName()),
+      new DatasetClass(AppWithTxAware.PedanticTxAware.class.getName())
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepositoryTest.java]: So this tests the cases of a dataset class that extends a class which implements Dataset, and the same for a static inner class. There is also the case of a class that implements an interface that extends Dataset. That is not captured by this test, but should be tested.

**Comment by anew** [comment]
I have just a few (mostly minor) comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java (hunk 4 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java:222]: this is done to get the application's config class and store the schema of the config. This was intended so that people could understand how to configure an application based on the restful call that gets application class details, though we haven't ended up using that at all.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java (hunk 1 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java (hunk 3 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java (hunk 2 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java]: The exception caught in line 81 is actually thrown by the `programRunnerFactory.create(ProgramType.SPARK)` call when Spark is not supported in the cluster.

If the `programRunner` is not a `ProgramClassLoaderProvider`, then the default program classloader will be used for the artifact inspection.

I've moved the `if` block outside of the `try` block to make it clearer.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java (hunk 7 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java:157]: yes and it includes non static ones as well. This is because every inner class in Java has it's own separate `.class` file accessible through the classpath.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java (hunk 4 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java:222]: It can be rewritten to use bytecode inspection as well, but probably would involve quite a lot of new code, mainly for resolving parameterized type recursively along the classes inheritance hierarchy.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java (hunk 5 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java]: updated.

File: cdap-app-fabric/src/test/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeServiceTest.java (hunk 6 above)
Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeServiceTest.java]: For the `ArtifactDetail` class, it is not. But in the context of this test, it doesn't matter as it is never used.
Updated the code to provide a dummy Location instead of `null` to safeguard against future changes.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepositoryTest.java (hunk 8 above)
Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepositoryTest.java]: Added a custom app to test various kinds of possible dataset class.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java (hunk 4 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java:222]: OK, let's do that in a separate PR/Jira

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java [hunk 9]
```diff
@@ -54,33 +54,33 @@
   }
 
   /**
-   * Create a classloader load classes from the given directory, with access to
+   * Create a classloader loads classes from a directory where an artifact jar has been expanded, with access to
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java]: that loads classes. (missing "that")

**Comment by anew** [comment]
Code looks good now and thanks for the good test coverage now! Only one more comment regarding a Javadoc. Please fix that and LGTM without further review.

**Comment by chtyim** [comment]
Addressed last comment. Squashed commits. Waiting for build at:

http://builds.cask.co/browse/CDAP-DUT4243

**Comment by chtyim** [comment]
Build passed
